pjsip/src/pjsip-simple/dlg_event.c

```diff
@@ -0,0 +1,509 @@
+    status = pjsip_evsub_create_uac(dlg,  &dlg_event_user, &STR_DIALOG,
+                                    options, &sub);
+    if (status != PJ_SUCCESS)
+        goto on_return;
+
+    /* Create dialog event */
+    dlgev = PJ_POOL_ZALLOC_T(dlg->pool, pjsip_dlg_event);
+    dlgev->dlg = dlg;
+    dlgev->sub = sub;
+    if (user_cb)
+        pj_memcpy(&dlgev->user_cb, user_cb, sizeof(pjsip_evsub_user));
+
+    pj_ansi_snprintf(obj_name, PJ_MAX_OBJ_NAME, "dlgev%p", dlg->pool);
+    dlgev->status_pool = pj_pool_create(dlg->pool->factory, obj_name,
+                                        512, 512, NULL);
+    pj_ansi_snprintf(obj_name, PJ_MAX_OBJ_NAME, "tmdlgev%p", dlg->pool);
+    dlgev->tmp_pool = pj_pool_create(dlg->pool->factory, obj_name,
+                                     512, 512, NULL);
+
+    /* Attach to evsub */
+    pjsip_evsub_set_mod_data(sub, mod_dlg_event.id, dlgev);
+
+    *p_evsub = sub;
+
+on_return:
+    pjsip_dlg_dec_lock(dlg);
+    return status;
+}
+
+
+/*
+ * Forcefully terminate dialog event.
+ */
+PJ_DEF(pj_status_t) pjsip_dlg_event_terminate(pjsip_evsub *sub,
+                                              pj_bool_t notify)
+{
+    return pjsip_evsub_terminate(sub, notify);
+}
+
+/*
+ * Create SUBSCRIBE
+ */
+PJ_DEF(pj_status_t) pjsip_dlg_event_initiate(pjsip_evsub *sub,
+                                             pj_int32_t expires,
+                                             pjsip_tx_data **p_tdata)
+{
+    return pjsip_evsub_initiate(sub, &pjsip_subscribe_method, expires,
+                                p_tdata);
+}
+
+
+/*
+ * Add custom headers.
+ */
+PJ_DEF(pj_status_t) pjsip_dlg_event_add_header(pjsip_evsub *sub,
+                                               const pjsip_hdr *hdr_list )
+{
+    return pjsip_evsub_add_header( sub, hdr_list );
+}
+
+
+/*
+ * Accept incoming subscription.
+ */
+PJ_DEF(pj_status_t) pjsip_dlg_event_accept(pjsip_evsub *sub,
+                                           pjsip_rx_data *rdata,
+                                           int st_code,
+                                           const pjsip_hdr *hdr_list)
+{
+    return pjsip_evsub_accept( sub, rdata, st_code, hdr_list );
+}
+
+
+/*
+ * Get dialog event status.
+ */
+PJ_DEF(pj_status_t) pjsip_dlg_event_get_status(pjsip_evsub *sub,
+                                               pjsip_dlg_event_status *status)
+{
+    pjsip_dlg_event *dlgev;
+
+    PJ_ASSERT_RETURN(sub && status, PJ_EINVAL);
+
+    dlgev = (pjsip_dlg_event*) pjsip_evsub_get_mod_data(sub, mod_dlg_event.id);
+    PJ_ASSERT_RETURN(dlgev!=NULL, PJSIP_SIMPLE_ENOPRESENCE);
+
+    if (dlgev->tmp_status._is_valid) {
+        PJ_ASSERT_RETURN(dlgev->tmp_pool!=NULL, PJSIP_SIMPLE_ENOPRESENCE);
+        pj_memcpy(status, &dlgev->tmp_status, sizeof(pjsip_dlg_event_status));
+    } else {
+        PJ_ASSERT_RETURN(dlgev->status_pool!=NULL, PJSIP_SIMPLE_ENOPRESENCE);
+        pj_memcpy(status, &dlgev->status, sizeof(pjsip_dlg_event_status));
+    }
```

Comment:
This is one possible race condition scenario:

1. Thread 1: executes `pjsip_dlg_event_get_status()`, `dlgev->tmp_status._is_valid` is evaluated as true, but before `memcpy(status, tmp_status, ..)`, context switch happens.
2. Thread 2: executes `dlg_event_process_rx_notify()` in this line:
    ```
    dlgev->tmp_status._is_valid = PJ_FALSE;
    pj_pool_reset(dlgev->tmp_pool);
    ```
3. Back to Thread 1: it copies from `dlgev->tmp_status` whose pool has just been reset (not swapped due to `p_st_code` is not 2xx).

This may be a corner case as usually NOTIFY will be responded with 200.

Reply:
OK, so how do we proceed here? I only add lock for dlg event and create the fix for presence in a separate PR?

Reply:
Yes, I think so.